src/dispatching/dialogue/storage/mod.rs

```diff
@@ -74,64 +74,59 @@ pub trait Storage<D> {
     ) -> BoxFuture<'static, Result<Option<D>, Self::Error>>;
 
     /// Erases [`Self::Error`] to [`std::error::Error`].
-    fn erase(self: Arc<Self>) -> ErasedStorage<D>
+    #[must_use]
+    fn erase(self: Arc<Self>) -> Arc<ErasedStorage<D>>
     where
         Self: Sized + Send + Sync + 'static,
-        Self::Error: std::error::Error + 'static,
+        Self::Error: std::error::Error + Send + Sync + 'static,
     {
-        struct Eraser<S>(Arc<S>);
-
-        impl<D, S> Storage<D> for Eraser<S>
-        where
-            S: Storage<D> + Send + Sync + 'static,
-            S::Error: std::error::Error + 'static,
-        {
-            type Error = Box<dyn std::error::Error>;
-
-            fn remove_dialogue(
-                self: Arc<Self>,
-                chat_id: i64,
-            ) -> BoxFuture<'static, Result<(), Self::Error>>
-            where
-                D: Send + 'static,
-            {
-                Box::pin(async move {
-                    Arc::clone(&self.0).remove_dialogue(chat_id).await.map_err(|e| e.into())
-                })
-            }
-
-            fn update_dialogue(
-                self: Arc<Self>,
-                chat_id: i64,
-                dialogue: D,
-            ) -> BoxFuture<'static, Result<(), Self::Error>>
-            where
-                D: Send + 'static,
-            {
-                Box::pin(async move {
-                    Arc::clone(&self.0)
-                        .update_dialogue(chat_id, dialogue)
-                        .await
-                        .map_err(|e| e.into())
-                })
-            }
-
-            fn get_dialogue(
-                self: Arc<Self>,
-                chat_id: i64,
-            ) -> BoxFuture<'static, Result<Option<D>, Self::Error>> {
-                Box::pin(async move {
-                    Arc::clone(&self.0).get_dialogue(chat_id).await.map_err(|e| e.into())
-                })
-            }
-        }
-
         Arc::new(Eraser(self))
     }
 }
 
-/// A storage with an erased error type.
-pub type ErasedStorage<D> = Arc<dyn Storage<D, Error = Box<dyn std::error::Error>>>;
+struct Eraser<S>(Arc<S>);
```

Comment:
What is the point in moving this?

Reply:
You mean extracting `struct Eraser` from the function? Not to pollute the trait definition too much, minus two nesting levels, which is easier for reading.